File: sdks/node/packages/sdk/src/runtime.ts
/**
 * FireweaveRuntime: shared engine behind the OpenFeature provider and the
 * FireweaveClient extensions. Owns lifecycle state machine, config validation,
 * context policy, and decision construction. Defaults never throw: evaluation
 * always returns a Decision.
 */
import { FireweaveError, isFireweaveError } from './errors.js';
import { assertHostAllowed } from './hosts.js';
import {
  DEFAULT_CONTEXT_LIMITS,
  DEFAULT_RESERVED_ATTRIBUTE_KEYS,
  mergeContexts,
  type ContextInput,
  type ContextLimits,
  type ContextPolicy,
} from './context.js';
import {
  matchesExpectedType,
  validateContext,
  validateControlPointKey,
  validateDefaultValue,
  type ExpectedFlagType,
} from './validation.js';
import type {
  AdapterResolution,
  BackendAdapter,
  RegisterTargetOptions,
  RegisterTargetResult,
} from './adapter.js';
import type {
  CanonicalContext,
  Decision,
  DecisionReason,
  JsonValue,
  LifecycleState,
} from './types.js';

export type { ExpectedFlagType } from './validation.js';

export interface FireweaveRuntimeConfig {
  /**
   * Fireweave project/runtime key (`project-api-key_…`). Adapters may also read
   * it from their own options or `FW_PROJECT_API_KEY`.
   */
  projectApiKey?: string;
  /** Backend host (must be http(s) and pass the allowlist). */
  host?: string;
  /**
   * SSRF allowlist: hostnames the SDK may talk to. Empty/undefined ⇒ the
   * canonical Fireweave default allowlist + loopback (DEFAULT_ALLOWED_HOSTS).
   * Custom/self-hosted endpoints must be listed explicitly; ['*'] opts out.
   */
  allowedHosts?: readonly string[];
  /** Require targetingKey on every evaluation. */
  requireTargetingKey?: boolean;
  /** Context bound overrides (defaults are the ratified spec bounds). */
  limits?: Partial<ContextLimits>;
  /** Additional reserved attribute keys. */
  reservedAttributeKeys?: readonly string[];
  /** Shutdown flush deadline (ms). Default DEFAULT_SHUTDOWN_TIMEOUT_MS. */
  shutdownTimeoutMs?: number;
}

/** Default bound on shutdown/flush (matches capabilities.get()). */
export const DEFAULT_SHUTDOWN_TIMEOUT_MS = 10000;

export interface EvaluateOptions {
  /** Attach flag payload as fireweave.payload metadata (sorted-key JSON string). */
  includePayload?: boolean;
  /**
   * When true, emit a Fireweave-owned exposure for a successful evaluation
   * (H-4 / ADR-0001 §23 errata / ruling 20). Default `false` — phase-one
   * evaluate is side-effect-free. Vendor `$feature_flag_called` stays
   * suppressed on the local snapshot path (RB-2); Fireweave owns emission
   * and dedup when opted in.
   */
  sendExposure?: boolean;
  signal?: AbortSignal;
}

function validateConfig(config: FireweaveRuntimeConfig): void {
  if (config.host !== undefined) {
    // Allowlist is ON by default (release-blockers H-1): undefined/empty
    // allowedHosts falls back to the canonical Fireweave + loopback list.
    assertHostAllowed(config.host, config.allowedHosts);
  }
  if (config.limits !== undefined) {
    for (const value of Object.values(config.limits)) {
      if (value !== undefined && (!Number.isInteger(value) || value <= 0)) {
        throw new FireweaveError('Configuration');
      }
    }
  }
}

/** JSON stringify with deterministic (sorted) key order at every level. */
export function stableStringify(value: JsonValue): string {
  if (value === null || typeof value !== 'object') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map((v) => stableStringify(v)).join(',')}]`;
  const keys = Object.keys(value).sort();
  return `{${keys.map((k) => `${JSON.stringify(k)}:${stableStringify(value[k] as JsonValue)}`).join(',')}}`;
}

export class FireweaveRuntime {
  readonly adapter: BackendAdapter;
  readonly config: FireweaveRuntimeConfig;
  private readonly contextPolicy: ContextPolicy;
  private state: LifecycleState = 'UNINITIALIZED';
  private globalContext: ContextInput | undefined;
  private clientContext: ContextInput | undefined;
  private initPromise: Promise<void> | undefined;
  private readonly stateListeners = new Set<(state: LifecycleState) => void>();
  /** Dedup window for evaluate/OF-path exposures (cleared on flush / shutdown). */
  private readonly evaluateExposureSeen = new Set<string>();

  constructor(adapter: BackendAdapter, config: FireweaveRuntimeConfig = {}) {
    this.adapter = adapter;
    this.config = { ...config };
    const limits: ContextLimits = { ...DEFAULT_CONTEXT_LIMITS, ...(config.limits ?? {}) };
    this.contextPolicy = {
      limits,
      reservedAttributeKeys: [
        ...DEFAULT_RESERVED_ATTRIBUTE_KEYS,
        ...(config.reservedAttributeKeys ?? []),
      ],
      requireTargetingKey: config.requireTargetingKey ?? false,
    };
  }

  /** Clear evaluate-path exposure dedup (M-2 / clear-on-flush lifecycle). */
  clearEvaluateExposureDedup(): void {
    this.evaluateExposureSeen.clear();
  }

  getState(): LifecycleState {
    return this.state;
  }

  onStateChange(listener: (state: LifecycleState) => void): void {
    this.stateListeners.add(listener);
  }

  private setState(next: LifecycleState): void {
    this.state = next;
    for (const listener of this.stateListeners) listener(next);
  }

  /** Mark definitions/cache staleness (adapter-driven). */
  markStale(): void {
    if (this.state === 'READY') this.setState('STALE');
  }

  setGlobalContext(context: ContextInput | undefined): void {
    this.globalContext = context;
  }

  setClientContext(context: ContextInput | undefined): void {
    this.clientContext = context;
  }

  /**
   * Initialize the runtime: validate config, then bring the adapter up.
   * Rejects with FireweaveError; Configuration errors put the runtime in FATAL.
   */
  async initialize(signal?: AbortSignal): Promise<void> {
    if (this.state === 'SHUTDOWN') {
      throw new FireweaveError('AlreadyClosed');
    }
    if (this.state === 'READY') return;
    if (this.initPromise !== undefined) return this.initPromise;

    this.initPromise = (async () => {
      this.setState('INITIALIZING');
      try {
        validateConfig(this.config);
      } catch (err) {
        this.setState('FATAL');
        throw err;
      }
      try {
        await this.adapter.initialize(signal);
        this.setState('READY');
      } catch (err) {
        if (isFireweaveError(err) && err.kind === 'Configuration') {
          this.setState('FATAL');
        } else {
          this.setState('ERROR');
        }
        throw isFireweaveError(err) ? err : new FireweaveError('Internal', { cause: err });
      } finally {
        this.initPromise = undefined;
      }
    })();
    return this.initPromise;
  }

  async shutdown(): Promise<void> {
    if (this.state === 'SHUTDOWN') return; // idempotent
    // The configured deadline bounds the whole flush+close sequence (M-1/L-4):
    // a wedged vendor client must not hang process exit.
    const timeoutMs = this.config.shutdownTimeoutMs ?? DEFAULT_SHUTDOWN_TIMEOUT_MS;
    let timer: ReturnType<typeof setTimeout> | undefined;
    const deadline = new Promise<void>((resolve) => {
      timer = setTimeout(resolve, timeoutMs);
    });
    const close = (async () => {
      try {
        await this.adapter.flush?.();
      } catch {
        // best-effort flush; never throw from shutdown
      }
      try {
        await this.adapter.shutdown();
      } catch {
        // swallow: shutdown must not throw
      }
    })();
    try {
      await Promise.race([close, deadline]);
    } finally {
      this.evaluateExposureSeen.clear();
      if (timer !== undefined) clearTimeout(timer);
      this.setState('SHUTDOWN');
    }
  }

  /** Canonicalize + validate the merged context (throws FireweaveError). */
  resolveContext(invocationContext: ContextInput | undefined): CanonicalContext {
    const merged = mergeContexts(this.globalContext, this.clientContext, invocationContext);
    const result = validateContext(merged, this.contextPolicy);
    if (!result.ok) throw result.error;
    return result.value;
  }

  /**
   * Evaluate a flag to a canonical Decision. Never throws.
   *
   * Validates in the fixed order spec/control-points.md "Validation, before
   * any I/O" names, stopping at the first failure: (1) key, (2) default vs
   * type, (3) context, (4) lifecycle. Only once all four pass does this
   * reach the adapter (the one I/O call in this method).
   */
  async evaluate(
    flagKey: string,
    expectedType: ExpectedFlagType,
    defaultValue: JsonValue,
    invocationContext?: ContextInput,
    options: EvaluateOptions = {},
  ): Promise<Decision> {
    const keyResult = validateControlPointKey(flagKey);
    if (!keyResult.ok) {
      return this.errorDecision(flagKey, defaultValue, keyResult.error);
    }

    const defaultResult = validateDefaultValue(expectedType, defaultValue);
    if (!defaultResult.ok) {
      return this.errorDecision(flagKey, defaultValue, defaultResult.error);
    }

    const merged = mergeContexts(this.globalContext, this.clientContext, invocationContext);
    const contextResult = validateContext(merged, this.contextPolicy);
    if (!contextResult.ok) {
      return this.errorDecision(flagKey, defaultValue, contextResult.error);
    }
    const context = contextResult.value;

    const lifecycleError = this.lifecycleError();
    if (lifecycleError !== undefined) {
      return this.errorDecision(flagKey, defaultValue, lifecycleError);
    }

    let resolution: AdapterResolution;
    try {
      const resolveOpts = options.signal !== undefined ? { signal: options.signal } : {};
      resolution = await this.adapter.resolve(flagKey, context, resolveOpts);
    } catch (err) {
      // H-2: non-Fireweave (vendor/internal) exception text never reaches the
      // outward errorMessage — the fixed taxonomy message is used and the
      // original error is preserved on `cause` only.
      const fw = isFireweaveError(err) ? err : new FireweaveError('Internal', { cause: err });
      return this.errorDecision(flagKey, defaultValue, fw);
    }

    if (!resolution.found) {
      // spec/modes.md "Behaviour per mode": local's unknown-key row is
      // `default` / `reason: DEFAULT` — deliberately not an error, unlike
      // remote's `default` / `ERROR` / `FlagNotFound`. An adapter signals the
      // former by carrying `reason: 'DEFAULT'` on its miss (FireweaveLocalAdapter);
      // any adapter that leaves `reason` unset (InMemoryAdapter,
      // FireweaveRemoteAdapter) keeps the FlagNotFound/ERROR path below.
      if (resolution.reason === 'DEFAULT') {
        return { flagKey, value: defaultValue, reason: 'DEFAULT', metadata: {} };
      }
      const meta: Record<string, string | number | boolean> = {};
      if (resolution.quotaLimited === true) meta['fireweave.quotaLimited'] = true;
      return this.errorDecision(flagKey, defaultValue, new FireweaveError('FlagNotFound', { metadata: meta }));
    }

    const value = resolution.value ?? null;
    if (!matchesExpectedType(value, expectedType)) {
      return this.errorDecision(flagKey, defaultValue, new FireweaveError('TypeMismatch'));
    }

    const metadata: Record<string, string | number | boolean> = {};
    if (resolution.version !== undefined) metadata['fireweave.flagVersion'] = resolution.version;
    // Detailed vendor fields travel together: they surface only when the
    // backend reported BOTH a vendor flag id and a matched condition index
    // (fixtures: eval-detailed-fields exposes them; eval-multivariate-string
    // [index, no id] and eval-payload-attached [id, no index] do not).
    if (resolution.vendorFlagId !== undefined && resolution.conditionIndex !== undefined) {
      metadata['fireweave.vendorFlagId'] = resolution.vendorFlagId;
      if (resolution.reasonCode !== undefined) metadata['fireweave.reasonCode'] = resolution.reasonCode;
    }
    if (options.includePayload === true && resolution.payload !== undefined) {
      metadata['fireweave.payload'] =
        typeof resolution.payload === 'string' ? resolution.payload : stableStringify(resolution.payload);
    }
    if (resolution.quotaLimited === true) metadata['fireweave.quotaLimited'] = true;
    if (resolution.fromCache === true) metadata['fireweave.fromCache'] = true;

    let reason: DecisionReason;
    if (resolution.enabled === false) {
      reason = 'DISABLED';
    } else if (resolution.reason !== undefined) {
      reason = resolution.reason;
    } else if (resolution.fromCache === true || this.state === 'STALE') {
      reason = 'STALE';
    } else {
      reason = 'TARGETING_MATCH';
    }

    const decision: Decision = { flagKey, value, reason, metadata };
    if (resolution.variant !== undefined) decision.variant = resolution.variant;

    // H-4 / ruling 20: evaluate is side-effect-free by default; opt in via sendExposure: true.
    if (options.sendExposure === true) {
      this.emitEvaluateExposure({
        targetingKey: context.targetingKey ?? '',
        flagKey,
        value,
        ...(resolution.variant !== undefined ? { variant: resolution.variant } : {}),
      });
    }

    return decision;
  }

  /**
   * Register a user or device so flag rules can target its DURABLE properties.
   *
   * Call once per login / device provisioning with the facts that outlive a
   * request (plan, beta membership, region, device model), then send the same
   * `targetingKey` on evaluation. Per-request context attributes still override
   * the registered properties for a single evaluation — the two identity paths
   * compose (spec/remote-protocol.md § Two identity paths).
   *
   * Resolves with `ok: false` instead of throwing: this runs in sign-in paths.
   * Adapters without the capability (in-memory, local dev) report
   * `Unsupported` so a dev harness does not silently look registered.
   */
  async registerTarget(
    targetingKey: string,
    options: RegisterTargetOptions = {},
  ): Promise<RegisterTargetResult> {
    const lifecycle = this.lifecycleError();
    if (lifecycle !== undefined) return { ok: false, error: lifecycle };
    if (this.adapter.registerTarget === undefined) {
      return { ok: false, error: new FireweaveError('UnsupportedCapability') };
    }
    return this.adapter.registerTarget(targetingKey, options);
  }

  private emitEvaluateExposure(exposure: {
    targetingKey: string;
    flagKey: string;
    value: JsonValue;
    variant?: string;
  }): void {
    if (exposure.targetingKey.length === 0) return;
    if (this.adapter.recordExposure === undefined) return;
    const key = `${exposure.targetingKey}\u0000${exposure.flagKey}\u0000${exposure.variant ?? ''}\u0000${stableStringify(exposure.value ?? null)}`;
    if (this.evaluateExposureSeen.has(key)) return;
    this.evaluateExposureSeen.add(key);
    this.adapter.recordExposure({
      targetingKey: exposure.targetingKey,
      flagKey: exposure.flagKey,
      value: exposure.value,
      ...(exposure.variant !== undefined ? { variant: exposure.variant } : {}),
    });
  }

  private lifecycleError(): FireweaveError | undefined {
    switch (this.state) {
      case 'UNINITIALIZED':
      case 'INITIALIZING':
        return new FireweaveError('NotReady');
      case 'ERROR':
        return new FireweaveError('NotReady');
      case 'FATAL':
        return new FireweaveError('Configuration');
      case 'SHUTDOWN':
        return new FireweaveError('AlreadyClosed');
      default:
        return undefined;
    }
  }

  private errorDecision(flagKey: string, defaultValue: JsonValue, err: FireweaveError): Decision {
    const metadata: Record<string, string | number | boolean> = {
      'fireweave.errorKind': err.kind,
      ...err.metadata,
    };
    return {
      flagKey,
      value: defaultValue,
      reason: 'ERROR',
      errorCode: err.openFeatureErrorCode,
      errorKind: err.kind,
      errorMessage: err.message,
      metadata,
    };
  }
}

File: sdks/python/src/fireweave/application/runtime.py
"""FireweaveRuntime: shared engine behind FireweaveClient. Owns lifecycle
state machine, context layering, and the evaluation pipeline. Reads never
raise: evaluation always returns a :class:`Decision`
(spec/control-points.md "Return discipline").
"""

from __future__ import annotations

import enum
import threading
from typing import Optional

from ..domain.context import ContextLimits, DEFAULT_RESERVED_ATTRIBUTE_KEYS, EvaluationContext, merge_contexts
from ..domain.decision import Decision, Reason
from ..domain.errors import (
    FLAG_METADATA_ERROR_KIND_KEY,
    AlreadyClosedError,
    ConfigurationError,
    FireweaveError,
    FlagNotFoundError,
    InternalError,
    NotReadyError,
    TypeMismatchError,
    UnsupportedCapabilityError,
)
from ..domain.types import FlagMetadata, FlagType
from ..domain.validation import (
    matches_expected_type,
    validate_context,
    validate_control_point_key,
    validate_default_value,
)
from .ports import BackendAdapter, FlagResolution, RegisterTargetOptions, RegisterTargetResult

__all__ = ["LifecycleState", "FireweaveRuntime", "DEFAULT_SHUTDOWN_TIMEOUT_MS"]

# Default bound on shutdown (matches node's DEFAULT_SHUTDOWN_TIMEOUT_MS).
DEFAULT_SHUTDOWN_TIMEOUT_MS = 10_000


class LifecycleState(enum.Enum):
    UNINITIALIZED = "UNINITIALIZED"
    INITIALIZING = "INITIALIZING"
    READY = "READY"
    STALE = "STALE"
    ERROR = "ERROR"
    FATAL = "FATAL"
    SHUTDOWN = "SHUTDOWN"

    @property
    def wire_name(self) -> str:
        """Provider-state name used by contracts fixtures."""
        return {
            LifecycleState.UNINITIALIZED: "NOT_READY",
            LifecycleState.INITIALIZING: "NOT_READY",
            LifecycleState.READY: "READY",
            LifecycleState.STALE: "STALE",
            LifecycleState.ERROR: "ERROR",
            LifecycleState.FATAL: "FATAL",
            LifecycleState.SHUTDOWN: "CLOSED",
        }[self]


class FireweaveRuntime:
    """Owns lifecycle, context layering, and the evaluation pipeline."""

    def __init__(
        self,
        adapter: BackendAdapter,
        *,
        limits: Optional[ContextLimits] = None,
        reserved_attribute_keys=(),
        require_targeting_key: bool = False,
        shutdown_timeout_ms: int = DEFAULT_SHUTDOWN_TIMEOUT_MS,
        global_context: Optional[EvaluationContext] = None,
    ) -> None:
        self._adapter = adapter
        self._limits = limits or ContextLimits()
        self._reserved_attribute_keys = tuple(DEFAULT_RESERVED_ATTRIBUTE_KEYS) + tuple(reserved_attribute_keys)
        self._require_targeting_key = require_targeting_key
        self._shutdown_timeout_ms = shutdown_timeout_ms
        self._global_context = global_context
        self._client_context: Optional[EvaluationContext] = None
        self._state = LifecycleState.UNINITIALIZED
        self._lock = threading.RLock()
        self._init_error: Optional[FireweaveError] = None

    # -- lifecycle -----------------------------------------------------------

    @property
    def state(self) -> LifecycleState:
        with self._lock:
            return self._state

    @property
    def adapter(self) -> BackendAdapter:
        return self._adapter

    def initialize(self) -> None:
        """Transition UNINITIALIZED -> READY; FATAL on configuration failure.

        Raises the underlying :class:`FireweaveError` so `init_fireweave`
        (application/mode.py) can propagate it — initialisation fails loudly
        (spec/modes.md); the runtime state is updated first so later
        evaluations degrade safely.
        """
        with self._lock:
            if self._state is LifecycleState.SHUTDOWN:
                raise AlreadyClosedError()
            if self._state is LifecycleState.READY:
                return
            self._state = LifecycleState.INITIALIZING
        try:
            self._adapter.initialize()
        except FireweaveError as exc:
            with self._lock:
                self._state = LifecycleState.FATAL
                self._init_error = exc
            raise
        except Exception as exc:  # vendor exception: wrap, preserve cause
            wrapped = InternalError("initialization failed")
            wrapped.__cause__ = exc
            with self._lock:
                self._state = LifecycleState.FATAL
                self._init_error = wrapped
            raise wrapped from exc
        with self._lock:
            self._state = LifecycleState.READY
            self._init_error = None

    def lifecycle_gate(self) -> Optional[FireweaveError]:
        """Extension-call lifecycle gate (kept for `invoke_capability`, even
        though SUPPORTED_CAPABILITIES is empty in v1 and never reaches it
        today): READY/STALE pass (``None``); after shutdown the gate is
        ``AlreadyClosed``; any pre-ready state degrades with
        ``UnsupportedCapability``. Callers convert the returned error into a
        structured result — extension APIs never raise for lifecycle
        reasons."""
        state = self.state
        if state in (LifecycleState.READY, LifecycleState.STALE):
            return None
        if state is LifecycleState.SHUTDOWN:
            return AlreadyClosedError()
        return UnsupportedCapabilityError()

    def mark_stale(self) -> None:
        with self._lock:
            if self._state is LifecycleState.READY:
                self._state = LifecycleState.STALE

    def force_state(self, state: LifecycleState) -> None:
        """Test hook: pin the lifecycle state."""
        with self._lock:
            self._state = state

    def shutdown(self) -> None:
        """Deterministic, idempotent shutdown; never raises."""
        with self._lock:
            if self._state is LifecycleState.SHUTDOWN:
                return
            self._state = LifecycleState.SHUTDOWN
        try:
            self._adapter.shutdown(self._shutdown_timeout_ms)
        except Exception:
            # Shutdown must not raise; adapter errors are swallowed by design.
            pass

    # -- context layering -----------------------------------------------------

    def set_global_context(self, context: Optional[EvaluationContext]) -> None:
        with self._lock:
            self._global_context = context

    def set_client_context(self, context: Optional[EvaluationContext]) -> None:
        with self._lock:
            self._client_context = context

    def merged_context(self, invocation: Optional[EvaluationContext]) -> EvaluationContext:
        with self._lock:
            global_ctx, client_ctx = self._global_context, self._client_context
        return merge_contexts(global_ctx, client_ctx, invocation)

    # -- target registration ---------------------------------------------------

    def register_target(
        self,
        targeting_key: str,
        options: Optional[RegisterTargetOptions] = None,
    ) -> RegisterTargetResult:
        """Register a user or device so rules can target its durable
        properties. Resolves ``ok=False`` instead of raising: this runs in
        sign-in paths, where a targeting concern must not break
        authentication (spec/modes.md "registerTarget in local mode").
        Adapters without the capability report `UnsupportedCapability`."""
        gate = self._lifecycle_error()
        if gate is not None:
            return RegisterTargetResult(ok=False, error=gate)
        register = getattr(self._adapter, "register_target", None)
        if not callable(register):
            return RegisterTargetResult(ok=False, error=UnsupportedCapabilityError())
        return register(targeting_key, options)

    def _lifecycle_error(self) -> Optional[FireweaveError]:
        """Evaluation/registration lifecycle gate (NotReady / AlreadyClosed)."""
        state = self.state
        if state in (LifecycleState.READY, LifecycleState.STALE):
            return None
        if state is LifecycleState.SHUTDOWN:
            return AlreadyClosedError()
        if state is LifecycleState.FATAL:
            return self._init_error or ConfigurationError()
        return NotReadyError()

    # -- evaluation pipeline -----------------------------------------------------

    def evaluate(
        self,
        flag_key: str,
        flag_type: FlagType,
        default_value,
        invocation_context: Optional[EvaluationContext] = None,
    ) -> Decision:
        """Evaluate a flag. Never raises; failures return the default.

        Validates in the fixed order spec/control-points.md "Validation,
        before any I/O" names, stopping at the first failure: (1) key, (2)
        default vs type, (3) context, (4) lifecycle. Only once all four pass
        does this reach the adapter (the one I/O call in this method).
        """
        key_result = validate_control_point_key(flag_key)
        if not key_result.ok:
            return self._error_decision(default_value, key_result.error)

        default_result = validate_default_value(flag_type, default_value)
        if not default_result.ok:
            return self._error_decision(default_value, default_result.error)

        merged = self.merged_context(invocation_context)
        context_result = validate_context(
            merged,
            limits=self._limits,
            reserved_keys=self._reserved_attribute_keys,
            require_targeting_key=self._require_targeting_key,
        )
        if not context_result.ok:
            return self._error_decision(default_value, context_result.error)
        canonical = context_result.value

        lifecycle_error = self._lifecycle_error()
        if lifecycle_error is not None:
            return self._error_decision(default_value, lifecycle_error)

        try:
            resolution = self._adapter.resolve(flag_key, canonical)
        except FireweaveError as exc:
            return self._error_decision(default_value, exc)
        except Exception as exc:
            wrapped = InternalError("evaluation failed")
            wrapped.__cause__ = exc
            return self._error_decision(default_value, wrapped)

        return self._decision_from_resolution(resolution, flag_type, default_value)

    # -- helpers ---------------------------------------------------------------

    def _decision_from_resolution(
        self,
        resolution: FlagResolution,
        flag_type: FlagType,
        default_value,
    ) -> Decision:
        if not resolution.matched:
            # spec/modes.md "Behaviour per mode": local's unknown-key row is
            # default/reason DEFAULT — deliberately not an error. Any adapter
            # that reports `matched=False` gets this branch (the strict
            # seam); an adapter signalling a genuine backend-side "unknown
            # key" instead RAISES FlagNotFoundError, which is caught above
            # and takes the ERROR branch below.
            return Decision(value=default_value, variant=None, reason=Reason.DEFAULT)

        if not matches_expected_type(resolution.value, flag_type):
            return self._error_decision(default_value, TypeMismatchError())

        value = resolution.value

        if resolution.fireweave_reason:
            reason = resolution.fireweave_reason
        elif not resolution.enabled:
            reason = Reason.DISABLED
        elif resolution.from_cache or self.state is LifecycleState.STALE:
            reason = Reason.STALE
        else:
            reason = Reason.TARGETING_MATCH

        metadata: FlagMetadata = {}
        if resolution.version is not None:
            metadata["fireweave.flagVersion"] = resolution.version
        # Detailed enrichment: only when the backend supplied a flag id, a
        # matched-condition index, AND a reason code together.
        if (
            resolution.vendor_flag_id is not None
            and resolution.condition_index is not None
            and resolution.reason_code is not None
        ):
            metadata["fireweave.vendorFlagId"] = resolution.vendor_flag_id
            metadata["fireweave.reasonCode"] = resolution.reason_code
        if resolution.from_cache:
            metadata["fireweave.fromCache"] = True
        metadata.update(resolution.extra_metadata)

        return Decision(value=value, variant=resolution.variant, reason=reason, flag_metadata=metadata)

    def _error_decision(self, default_value, error: FireweaveError) -> Decision:
        metadata: FlagMetadata = {FLAG_METADATA_ERROR_KIND_KEY: error.kind.value}
        if isinstance(error, FlagNotFoundError) and error.quota_limited:
            metadata["fireweave.quotaLimited"] = True
        return Decision(
            value=default_value,
            variant=None,
            reason=Reason.ERROR,
            error_code=error.openfeature_error_code,
            error_message=error.message,
            error_kind=error.kind,
            flag_metadata=metadata,
        )
